### ISSUE TITLE
False-positive type error using TypeVar with "is None" conditional

### ISSUE THREAD

**Issue by tpajenkamp-dspace** [microsoft/pyright#9456]
**Describe the bug**
Pyright reports an error for functions like the one below which uses a `TypeVar` for both an argument and its return type where `None` is one of the constraints. The type of the type variable is not properly narrowed (or whatever the correct term is in this case) by the `arg is None` conditional. If I use `isinstance(arg, types.NoneType)` instead it works, but this form does not seem idiomatic to me.

**Code**

```python
import typing

OptionalStr = typing.TypeVar("OptionalStr", str, None)

def string_fun(arg: OptionalStr) -> OptionalStr:
    typing.reveal_type(arg)     # Type of "arg" is "OptionalStr@string_fun"
    if arg is None:
        typing.reveal_type(arg) # Type of "arg" is "None*"
        return None             # error: Type "None" is not assignable to type "OptionalStr@string_fun" (reportReturnType)

    typing.reveal_type(arg)     # Type of "arg" is "str*"
    return f"s: {arg}"          # error: Type "str" is not assignable to type "OptionalStr@string_fun" (reportReturnType)

```

**VS Code extension or command-line**
pyright 1.1.388
`& 'C:\Python312\python.exe' -m pyright .`
